### PR TITLE
BB-804 - Clean up old MySQL instances as part of CI cleanup

### PR DESCRIPTION
**Pull request by jamestait**
This PR adds an extra cleanup instance to the Integration Cleanup script, to remove MySQL databases that are older than three days.

## Testing

You can test the cleanup function on a local OCIM instance:

    $ vagrant ssh
    (opencraft) vagrant@ubuntu-xenial:~$ cd opencraft
    (opencraft) vagrant@ubuntu-xenial:~/opencraft$ mysql -u root -h localhost
    Welcome to the MySQL monitor.  Commands end with ; or \g.
    Your MySQL connection id is 20                                                                                                                      
    Server version: 5.7.24-0ubuntu0.16.04.1 (Ubuntu)
    
    Copyright (c) 2000, 2018, Oracle and/or its affiliates. All rights reserved.
    
    Oracle is a registered trademark of Oracle Corporation and/or its
    affiliates. Other names may be trademarks of their respective
    owners.
    
    Type 'help;' or '\h' for help. Type '\c' to clear the current input statement.
    
    mysql> CREATE DATABASE f4ad7be3_integration_plebia_net_edxapp;
    Query OK, 1 row affected (0.01 sec)
    
    mysql> use f4ad7be3_integration_plebia_net_edxapp
    Database changed
    mysql> CREATE TABLE foo (id int);
    Query OK, 0 rows affected (0.04 sec)
    
    mysql> CREATE DATABASE unrelated)db;
    Query OK, 1 row affected (0.01 sec)
    
    mysql> use unrelated_db
    Database changed
    mysql> CREATE TABLE foo (id int);
    Query OK, 0 rows affected (0.04 sec)
    
    mysql> \q
    Bye
    (opencraft) vagrant@ubuntu-xenial:~/opencraft$ make shell
    ...
    In [1]: from cleanup_utils import mysql_cleanup
    
    In [2]: my_cleanup = mysql_cleanup.MySqlCleanupInstance(0, settings.DEFAULT_INSTANCE_MYSQL_URL, 'plebia.net', False)
    
    In [3]: my_cleanup._get_old_databases()
    Out[3]:
    (('f4ad7be3_integration_plebia_net_edxapp',
      datetime.datetime(2019, 1, 18, 11, 23, 58)),)
    
    In [4]: my_cleanup.run_cleanup()
    
    In [5]: my_cleanup._get_old_databases()
    Out[5]: ()
    
    In [6]: my_cleanup.cleaned_up_hashes
    Out[6]: ['f4ad7be3']
    
    In [7]: 
    Do you really want to exit ([y]/n)? 
    (opencraft) vagrant@ubuntu-xenial:~/opencraft$ mysql -u root -h localhost
    mysql> show DATABASES;
    +-------------------------------------------------+
    | Database                                        |
    +-------------------------------------------------+
    | information_schema                              |
    | ...                                             |
    | mysql                                           |
    | performance_schema                              |
    | sys                                             |
    | unrelated_db                                    |
    +-------------------------------------------------+
    38 rows in set (0.00 sec)

    mysql>

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -49,13 +49,13 @@ Before activating a server, there's the option to test it through a
 basic-auth password-protected link in the "Authenticated Link" section
 (the username and password are embedded in the link).
 
-Sometimes Open edX playbook fails, and then you need to read the log,
```
Comment (giovannicimolin) [inline README.md]: These changes are outside the scope of BB-804. 
Please remove them from this PR.

File: cleanup_utils/mysql_cleanup.py [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+    """
+    Handles the cleanup of old MySQL databases
+    """
+    def __init__(self, age_limit, url, domain, dry_run):
+        """
+        Set up variables needed for cleanup
+        """
+        self.age_limit = age_limit
+        self.cleaned_up_hashes = []
+        self.cursor = _get_cursor(url)
```
Comment (giovannicimolin) [inline cleanup_utils/mysql_cleanup.py]: If we pass a URL that is either broke or unaccessible, this will throw an exception. 
Can you add a handler for this to skip cleaning up MySQL if a connection error happens?
I've done something similar [here](https://github.com/open-craft/opencraft/blob/master/cleanup_utils/openstack_cleanup.py#L51).

We want this script to be as error tolerant as possible. This is needed because if this module throws an exception, we will lose all those hashes from the cleanups that ran before this one.

File: cleanup_utils/mysql_cleanup.py [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+    database_url_obj = urlparse(url)
+    try:
+        connection = mysql.connect(
+            host=database_url_obj.hostname,
+            user=database_url_obj.username or '',
+            passwd=database_url_obj.password or '',
+            port=database_url_obj.port or 3306,
+        )
+    except MySQLError as exc:
+        logger.excecption('Cannot get MySQL cursor: %s', exc)
```
Comment (giovannicimolin) [inline cleanup_utils/mysql_cleanup.py]: There's a typo here: `excecption`

File: cleanup_utils/mysql_cleanup.py [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+                )
+                continue
+
+            logger.info(
+                '    * Dropping MySQL DB `%s` created at %s', database,
+                datetime.strftime(create_date, '%Y-%m-%dT%H:%M:%SZ'))
+            if not self.dry_run:
+                try:
+                    self.cursor.execute(
+                        'DROP DATABASE IF EXISTS {}'.format(database))
```
Comment (giovannicimolin) [inline cleanup_utils/mysql_cleanup.py]: Nit: Add a line break on the parenthesis.

File: cleanup_utils/mysql_cleanup.py [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+                    self.cleaned_up_hashes.append(match.groups()[0])
+                except MySQLError as exc:
+                    logger.exception(
+                        'Unable to remove MySQL DB: %s. %s', database, exc)
```
Comment (giovannicimolin) [inline cleanup_utils/mysql_cleanup.py]: Nit: Add a line break on the parenthesis.

File: cleanup_utils/mysql_cleanup.py [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+        if self.dry_run:
+            logger.info("Running in DRY_RUN mode, no actions will be taken.")
+
+        databases = self._get_old_databases()
+        logger.info('Found %d old databases', len(databases))
+
+        for (database, create_date) in databases:
+            logger.info('  > Considering database %s', database)
+            instance_database_re = r'^([0-9a-f]{8})([_0-9a-z]+)?_%s' % (
+                self.domain_suffix,)
```
Comment (giovannicimolin) [inline cleanup_utils/mysql_cleanup.py]: You don't need this line break here, this will improve readability.

File: cleanup_utils/integration_cleanup.py [hunk 7]
```diff
@@ -31,13 +31,17 @@
...
+from cleanup_utils.openstack_cleanup import OpenStackCleanupInstance
 
 
 # Constants ###################################################################
 
-DEFAULT_AGE_LIMIT = datetime.utcnow().replace(tzinfo=UTC) - timedelta(days=3)
+# Default age at which things should be cleaned up, in days
+DEFAULT_AGE_LIMIT = 3
+DEFAULT_CUTOFF_TIME = (
+    datetime.utcnow().replace(tzinfo=UTC) - timedelta(days=DEFAULT_AGE_LIMIT))
```
Comment (lgp171188) [inline cleanup_utils/integration_cleanup.py]: @jamestait, don't you think the below style of brackets is more readable and looks cleaner?

```python
DEFAULT_CUTOFF_TIME = (
    datetime.utcnow().replace(tzinfo=UTC) - timedelta(days=DEFAULT_AGE_LIMIT)
)
```

File: cleanup_utils/integration_cleanup.py [hunk 8]
```diff
@@ -103,7 +116,9 @@ def run_integration_cleanup(dry_run=False):
     )
     # Run DNS cleanup erasing all integration entries except for those on
     # the deletion_blacklist
-    hashes_to_clean = aws_cleanup.cleaned_up_hashes + os_cleanup.cleaned_up_hashes
+    hashes_to_clean = (
+        aws_cleanup.cleaned_up_hashes + os_cleanup.cleaned_up_hashes +
+        mysql_cleanup.cleaned_up_hashes)
```
Comment (lgp171188) [inline cleanup_utils/integration_cleanup.py]: @jamestait, similar to a previous comment, it is a good idea to move the closing parenthesis to a new line.

```python
hashes_to_clean = (
    aws_cleanup.cleaned_up_hashes + os_cleanup.cleaned_up_hashes +
    mysql_cleanup.cleaned_up_hashes
)
```

File: cleanup_utils/mysql_cleanup.py [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+        query = """SELECT table_schema, MAX(create_time) AS create_time
+            FROM information_schema.tables
+            WHERE create_time <= DATE_SUB(NOW(), INTERVAL %(age_limit)s DAY)
+            AND table_schema LIKE %(domain_filter)s
+            GROUP BY table_schema ORDER BY create_time DESC"""
+        params = {
+            'age_limit': self.age_limit,
+            'domain_filter': '%_{}_%'.format(self.domain_suffix)
+        }
+        self.cursor.execute(query, params)
```
Comment (lgp171188) [inline cleanup_utils/mysql_cleanup.py]: @jamestait, you have to wrap this in a `try-except` block and then log an error message if something goes wrong here.

File: cleanup_utils/mysql_cleanup.py [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+
+        if self.dry_run:
+            logger.info("Running in DRY_RUN mode, no actions will be taken.")
+
+        databases = self._get_old_databases()
+        logger.info('Found %d old databases', len(databases))
+
+        for (database, create_date) in databases:
+            logger.info('  > Considering database %s', database)
+            instance_database_re = r'^([0-9a-f]{8})([_0-9a-z]+)?_%s' % (self.domain_suffix,)
```
Comment (lgp171188) [inline cleanup_utils/mysql_cleanup.py]: @jamestait, this may not be required because the databases are already filtered using the base domain name and it is unlikely to have a database on this RDS server which matches the domain name but doesn't match the pattern. All the databases in this MySQL server are used solely for the CI checks.

File: cleanup_utils/mysql_cleanup.py [hunk 11]
```diff
@@ -0,0 +1,136 @@
...
+                    instance_database_re
+                )
+                continue
+
+            logger.info(
+                '    * Dropping MySQL DB `%s` created at %s', database,
+                datetime.strftime(create_date, '%Y-%m-%dT%H:%M:%SZ'))
+            if not self.dry_run:
+                try:
+                    self.cursor.execute(
```
Comment (lgp171188) [inline cleanup_utils/mysql_cleanup.py]: @jamestait, I am wondering if we need to do this one by one instead of using `executemany()` to do it in one shot. I know that the performance of the `executemany()` method varies from driver to driver and also based on the backend, but can you check and find out if it makes sense to use it?

Comment (jamestait) [reply cleanup_utils/mysql_cleanup.py]: From the docstring for `MySQLdb.cursors.executemany()`:

    """Execute a multi-row query.

    :param query: query to execute on server
    :param args:  Sequence of sequences or mappings.  It is used as parameter.
    :return: Number of rows affected, if any.

    This method improves performance on multiple-row INSERT and
    REPLACE. Otherwise it is equivalent to looping over args with
    execute().
    """

So I don't think we gain anything with `executemany()` but we'd lose the ability to report individually on each operation.

File: cleanup_utils/mysql_cleanup.py (hunk 10 above)
Comment (jamestait) [reply cleanup_utils/mysql_cleanup.py]: This is serving a dual purpose: it's an additional precaution against dropping unrelated databases, but it's also extracting the hash to return to the main script so it can be used to clean up old DNS entries. We could, perhaps, go ahead with the removal even if the RE doesn't match, in which case no hash would be returned for that database, but I've erred on the side of caution in case this is used elsewhere, where our assumptions don't hold.

Comment (lgp171188) [reply cleanup_utils/mysql_cleanup.py]: @jamestait, it will be good to mention the gist of this explanation in a comment before this block of code.

File: cleanup_utils/integration_cleanup.py (hunk 7 above)
Comment (jamestait) [reply cleanup_utils/integration_cleanup.py]: Yes, thank you - fixed this and the other instances. :)

File: cleanup_utils/mysql_cleanup.py (hunk 9 above)
Comment (jamestait) [reply cleanup_utils/mysql_cleanup.py]: Done. In addition, I've returned an empty list in that instance, so the remainder of the cleanup will complete normally - we'll log the exception, log that no databases will be cleaned up, and the DNS cleanup will proceed as normal.

File: cleanup_utils/mysql_cleanup.py (hunk 10 above)
Comment (jamestait) [reply cleanup_utils/mysql_cleanup.py]: You read my mind - added. :)

File: README.md (hunk 1 above)
Comment (giovannicimolin) [reply README.md]: Ah, I've just asked because you also updated this block of text here, which isn't related to the cleanup.
```
Sometimes Open edX playbook fails, and then you need to read the log, which is
shown in real-time in the web console.  You can fix the settings and then spawn
another server.  Failed and old inactive servers and MySQL databases are
automatically cleaned up after three days.  An important feature is that Ocim
*grants SSH access* to members of a configurable GitHub organization, so you
can always SSH to an appserver's IP, *even if Open edX's deployment failed*,
and then debug it.  You can use your GitHub username and key.
```

Comment (jamestait) [reply README.md]: There are changes that are related to the PR:

 * "old inactive server **and MySQL databases** are"
 * "automatically cleaned up after **three days**" rather than "**some configurable amount of days**".

But then I re-flowed the text for better readability, which isn't related.

Would you prefer if I reverted these changes and made them in a separate PR?

Comment (giovannicimolin) [reply README.md]: @jamestait This paragraph is talking about Ocim automatic cleanup of Appservers and isn't related to the integration cleanup.
If you leave an old Appserver or a deployment through Ocim fails, it'll be automatically cleaned after 3 days, and this process isn't related to the CI runs :grin:

Comment (jamestait) [reply README.md]: Ah, got it! I thought it was actually the CI cleanup process that was responsible for this, sorry! In that case, I'll revert this change.

**Comment by giovannicimolin** [comment]
@jamestait I think you are only missing the MySQL requirement on `cleanup_utils/requirements.txt`.

**Comment by jamestait** [comment]
Thanks for the nudge, @giovannicimolin - I've added the requirement. :smiley_cat:

File: cleanup_utils/integration_cleanup.py [hunk 12]
```diff
@@ -89,12 +94,21 @@ def run_integration_cleanup(dry_run=False):
...
         dry_run=dry_run
     )
     os_cleanup.run_cleanup()
 
+    # Run MySQL cleanup
+    mysql_cleanup = MySqlCleanupInstance(
+        age_limit=DEFAULT_AGE_LIMIT,
+        url=os.environ['DEFAULT_INSTANCE_MYSQL_URL'],
+        domain=os.environ['DEFAULT_INSTANCE_BASE_DOMAIN'],
+        dry_run=args.dry_run
```
Comment (lgp171188) [inline cleanup_utils/integration_cleanup.py]: @jamestait, there is no variable `args` in this method - it was removed in @giovannicimolin's recent PR which you had to rebase your changes on. So how did this even work? 🤔

```suggestion
        dry_run=dry_run
```

File: cleanup_utils/mysql_cleanup.py [hunk 13]
```diff
@@ -0,0 +1,146 @@
...
+            FROM information_schema.tables
+            WHERE create_time <= DATE_SUB(NOW(), INTERVAL %(age_limit)s DAY)
+            AND table_schema LIKE %(domain_filter)s
+            GROUP BY table_schema ORDER BY create_time DESC"""
+        params = {
+            'age_limit': self.age_limit,
+            'domain_filter': '%_{}_%'.format(self.domain_suffix)
+        }
+        try:
+            self.cursor.execute(query, params)
```
Comment (lgp171188) [inline cleanup_utils/mysql_cleanup.py]: @jamestait, wouldn't this statement throw an unhandled `AttributeError: 'NoneType' object has no attribute 'execute'` exception when the code to create a connection and then retrieve the cursor throws an exception, thereby causing the `self.cursor` value to be set to `None`?

File: cleanup_utils/integration_cleanup.py (hunk 12 above)
Comment (giovannicimolin) [reply cleanup_utils/integration_cleanup.py]: @lgp171188 As the args definition happens outside a function, it becomes a global variable. Nevertheless, I've created `run_integration_cleanup` method so we can run the cleanup from python shell if needed, and this `args.dry_run` would throw an error, so it must be changed to the parameter `dry_run` as suggested.

Comment (jamestait) [reply cleanup_utils/integration_cleanup.py]: Great catch, thanks! Fixed.

File: cleanup_utils/mysql_cleanup.py (hunk 13 above)
Comment (jamestait) [reply cleanup_utils/mysql_cleanup.py]: Indeed it would. This wouldn't be a problem in the normal case, as `run_cleanup` would have already caught that and skipped the cleanup, but this method should be robust against such a case. I've moved the check to here, and just return an empty list in the event the cursor is `None`.
